fuzzers/main.py

```diff
@@ -0,0 +1,642 @@
+            if artifact in REPORTED_ARTIFACTS[other_branch]:
+                already_reported_for.append(other_branch)
+        REPORTED_ARTIFACTS[branch].append(artifact)
+
+        if len(already_reported_for) == 0:
+            already_reported_msg = ""
+        else:
+            already_reported_msg = f"(already reported for branches {already_reported_for})"
+
+        # Send the information in two mesages, a short one guaranteed to succeed, then a long one
+        # with the log lines that could go over the message size limit
+        import textwrap
+        import zulip
+        client = zulip.Client(config_file="~/.fuzzer-zuliprc")
+        client.send_message({
+            "type": "stream",
+            "to": "nearinc/fuzzer/private",
+            "topic": f"{branch}: artifact {artifact}",
+            "content": (
+                f"# Fuzzer found new crash for branch *{branch}* {already_reported_msg}\n"
+                f"Full logs are available at {logs_url}.\n"
+                f"\n"
+                f"You can download the artifact by using the following command (all commands are to be run from the root of `nearcore`):\n"
+                f"```\n"
+                f"{downloader}\n"
+                f"```\n"
+                f"\n"
+                f"Then, you can reproduce by running the following command:\n"
+                f"```\n"
+                f"{reproducer}\n"
+                f"```\n"
+                f"\n"
+                f"Or minimize by running the following command:\n"
+                f"```\n"
+                f"{minimizer}\n"
+                f"```\n"
+                f"\n"
+                f"Please edit the topic name to add more meaningful information once investigated. Keeping the artifact hash in it can help if the same artifact gets detected as crashing another branch.\n"
+            ),
+        })
+        last_log_lines = ""
+        for l in log_lines[::-1]:
+            if l.startswith("```"):
+                # Censor the end of a spoiler block, not great but this is for human consumption
+                # anyway
+                l = " " + l
+            if len(last_log_lines) + len(l) > 9000:
+                # Zulip limit is 10k, let's keep some safety buffer here
+                break
+            last_log_lines = l + last_log_lines
+        client.send_message({
+            "type": "stream",
+            "to": "nearinc/fuzzer/private",
+            "topic": f"{branch}: artifact {artifact}",
+            "content": f"```spoiler Last few log lines\n{last_log_lines}\n```",
+        })
+
+    def signal(self, signal: int) -> None:
+        """
+        Signal the fuzzer process (with all its process group as a fuzzer can spawn sub-fuzzers)
+        """
+
+        print(f"Sending signal {signal} to fuzzer {self.proc.pid}")
+        os.killpg(os.getpgid(self.proc.pid), signal)
+
+def random_weighted(array, num: int):
+    return random.choices(array, [x['weight'] for x in array], k = num)
```

Comment:
```
def random_weighted(array: typing.Sequence[dict], num: int) -> dict:
```

would improve typing slightly

Reply:
Hmm so the latest version already has something better than the lines you quoted, so I think it's already good-ish?